patches/api/0440-More-PortalCreateEvent-reasons.patch

```diff
@@ -0,0 +1,72 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Sun, 16 May 2021 13:40:23 -0700
+Subject: [PATCH] More PortalCreateEvent reasons
+
+
+diff --git a/src/main/java/org/bukkit/event/world/PortalCreateEvent.java b/src/main/java/org/bukkit/event/world/PortalCreateEvent.java
+index 579f017474ff22f0991ca884c35cdde7e14a94dc..1b24ca4020373bb5d66f900c7a0a82609568ac30 100644
+--- a/src/main/java/org/bukkit/event/world/PortalCreateEvent.java
++++ b/src/main/java/org/bukkit/event/world/PortalCreateEvent.java
+@@ -11,6 +11,8 @@ import org.jetbrains.annotations.Nullable;
+ 
+ /**
+  * Called when a portal is created
++ * <p>
++ * May be trigger asynchronously for world generation-related reasons.
+  */
+ public class PortalCreateEvent extends WorldEvent implements Cancellable {
+     private static final HandlerList handlers = new HandlerList();
+@@ -25,7 +27,12 @@ public class PortalCreateEvent extends WorldEvent implements Cancellable {
+     }
+ 
+     public PortalCreateEvent(@NotNull final List<BlockState> blocks, @NotNull final World world, @Nullable Entity entity, @NotNull CreateReason reason) {
+-        super(world);
++        // Paper start
++        this(false, blocks, world, entity, reason);
++    }
++    public PortalCreateEvent(final boolean async, @NotNull final List<BlockState> blocks, @NotNull final World world, @Nullable Entity entity, @NotNull CreateReason reason) {
++        super(world, async);
++        // Paper end
+ 
+         this.blocks = blocks;
+         this.entity = entity;
+@@ -101,6 +108,37 @@ public class PortalCreateEvent extends WorldEvent implements Cancellable {
+          * When the target end platform is created as a result of a player
+          * entering an end portal.
+          */
+-        END_PLATFORM
++        END_PLATFORM,
++        // Paper start
++        /**
++         * When the ender dragon is defeated and a new end gateway
++         * is generated on the main end island.
++         */
++        END_GATEWAY_DRAGON_DEFEAT,
++
++        /**
++         * When an entity causes the generation of an end gateway
++         * by going through one that does not have a linked gateway.
++         */
++        END_GATEWAY_PAIR,
++
++        /**
++         * When an inactive end portal is generated on the main end
++         * island in the end dimension. This can happen in normal world generation or
++         * when a new dragon is spawned, replacing the active portal with a inactive one.
++         */
++        END_PORTAL,
++
++        /**
++         * When an active end portal is generated on the main end
++         * island in the end dimension.
++         */
++        ACTIVE_END_PORTAL,
```

Comment:
Might be clearer if we use EXIT_END_PORTAL (or END_EXIT_PORTAL) here to avoid confusion with activating the portal in the overworld.